### ISSUE TITLE
Cache generated textures for SpriteCircle and SpriteSolidColor

### ISSUE THREAD

**Issue by pushfoo** [pythonarcade/arcade#973]
### What should be added/changed?

**tl;dr** [`arcade.SpriteSolidColor`](https://github.com/pythonarcade/arcade/blob/d5ef3c5cd187efe199b844da9977bb3062a41383/arcade/sprite.py#L1262) and [`arcade.SpriteCircle`](https://github.com/pythonarcade/arcade/blob/d5ef3c5cd187efe199b844da9977bb3062a41383/arcade/sprite.py#L1283) should cache the auto-generated textures for each combination of constructor arguments they recieve.

Currently, instantiating either class generates an entirely new copy of a texture using PIL. This can cause slowdowns when many instances of `SpriteCircle` or subclasses are instantiated because PIL uses CPU-bound drawing to generate the textures. This is especially evident at larger sizes for soft circles. 

This example demonstrates the texture growth:
```py
from collections import Set

import arcade


class TextureAllocationExample(arcade.Window):
    def __init__(self, width, height, title):
        super().__init__(width, height, title)
        self.sprites: arcade.SpriteList = arcade.SpriteList()

        # track individual ID numbers
        self.tex_id_numbers: Set[int] = set()
        self.tex_pil_image_id_numbers: Set[int] = set()
        # track texture names
        self.tex_names: Set[str] = set()

    def on_update(self, delta_time):
        sprite = arcade.SpriteCircle(5, arcade.color.GREEN)

        self.tex_id_numbers.add(id(sprite))
        self.tex_pil_image_id_numbers.add(id(sprite.texture.image))
        self.tex_names.add(sprite.texture.name)

        self.sprites.append(sprite)
       
        # repeated copies of the same texture are allocated
        print(
            f" tex_id_numbers {len(self.tex_id_numbers)}"
            f" tex_pil_image_id_numbers {len(self.tex_pil_image_id_numbers)}"
            f" tex_names {len(self.tex_names)}"
        )

if __name__ == "__main__":
    window = TextureAllocationExample(640, 480, "Texture allocation example")
    arcade.run()
```

Instead, these textures should be cached and re-used for instantiations of the class that have the same parameters. A final fix could use the global texture cache in `arcade/texture.py`, or a local per-class cache like this workaround I have in my prototype:
```py
class CachedSpriteCircle(arcade.Sprite):

    # quick fix until fix is made for libraries
    PARAM_CACHE: Dict[Tuple, arcade.Texture] = {}

    def __init__(self, radius, color: Color):
        super().__init__()
        diameter = int(radius * 2)
        params = (diameter, color)
        if params not in self.PARAM_CACHE:
            self.PARAM_CACHE[params] = make_circle_texture(diameter, color)

        self.radius = radius
        self.base_color = color

        self.texture = self.PARAM_CACHE[params]
        self._points = self.texture.hit_box_points
```

If a quick fix to use caches to limit PIL calls is ok, I can take this ticket.

### What would it help with?

**tl;dr** improve performance and instill good game prototyping habits in students

#### Performance

Early on in the development process, people often want to prototype gameplay without concern for assets. I did this myself recently when I used SpriteCircle to create asset stand-in projectiles and enemies for a pymunk based prototype. That lead to occasional slowdowns when I fired multiple projectiles per second or spawned many enemies in a single frame. 

#### Build Focus, Good Habit, and Project Momentum

As Arcade aims to be an education-friendly framework, it would be a good idea to instill good game prototyping habits into students early. One common problem students and other aspiring developers frequently run into is that they focus on assets before gameplay, and get overwhelmed or miss deadlines as they try to tweak the perfect assets into being. Performance issues early on may also discourage new developers. Improving the performance of these classes will help students prototype effectively and keep project momentum high.
